src/components/Modal/ModalSelectAssets.tsx

```diff
@@ -35,51 +41,83 @@ export const ModalSelectAssets = () => {
 
   const handleSearchClear = () => setSearchValue("")
 
-  const filterPattern = (asset: BaseTokenInfo) =>
-    asset.name
+  const filterPattern = (asset: SelectItemToken) =>
+    asset.token.name
       .toLocaleUpperCase()
       .includes(deferredQuery.toLocaleUpperCase()) ||
-    asset.chainName
-      .toLocaleUpperCase()
-      .includes(deferredQuery.toLocaleUpperCase())
+    (isBaseToken(asset.token)
+      ? asset.token.chainName
+          .toLocaleUpperCase()
+          .includes(deferredQuery.toLocaleUpperCase())
+      : true)
+
+  const handleSelectToken = (token: SelectItemToken) => {
+    if (modalType !== ModalType.MODAL_SELECT_ASSETS) {
+      throw new Error("Invalid modal type")
+    }
 
-  const handleSelectToken = (token: BaseTokenInfo) => {
-    onCloseModal({
-      ...(payload as { fieldName: string }),
-      modalType,
-      token,
-    })
+    const newPayload: ModalSelectAssetsPayload = {
+      ...(payload as ModalSelectAssetsPayload),
+      modalType: ModalType.MODAL_SELECT_ASSETS,
+      token: token.token,
+    }
+    onCloseModal(newPayload)
   }
 
   useEffect(() => {
     if (!data.size && !isLoading) {
       return
     }
     const { selectToken, fieldName } = payload as {
-      selectToken: NetworkTokenWithSwapRoute | undefined
+      selectToken: Token | undefined
       fieldName: string
     }
 
-    const getAssetList: TokenListWithNotSelectableToken[] = []
-    const getAssetListWithBalances: TokenListWithNotSelectableToken[] = []
-    for (const value of data.values()) {
-      let isNotSelectable = false
+    const selectedTokenId = selectToken
+      ? isBaseToken(selectToken)
+        ? selectToken.defuseAssetId
+        : selectToken.unifiedAssetId
+      : undefined
+
+    const getAssetList: SelectItemToken[] = []
+    const getAssetListWithBalances: SelectItemToken[] = []
+    for (const [tokenId, token] of data) {
       // We do not filter "tokenIn" as give full access to tokens in first step
       // Filtration by routes should happen only at "tokenOut"
-      const isAlreadySelected =
-        selectToken && value.defuseAssetId === selectToken.defuseAssetId
-      if (isAlreadySelected) {
-        isNotSelectable = true
-      }
+      const disabled = selectedTokenId != null && tokenId === selectedTokenId
 
-      if (value?.balance) {
-        getAssetListWithBalances.push({
-          ...value,
-          balance: value?.balance,
-          isNotSelectable,
-        })
+      if (isBaseToken(token)) {
+        if (
+          token.balance != null &&
```

Comment:
This if condition is incorrect because balance and balanceUsd (calculated using convertedLast, the token rate) represent different features. Even if we successfully fetch the token balance (used for input/output amounts in the swap, deposit, or withdraw forms), the API request to get the market values (for USD conversion) could still fail. Therefore, displaying the token balance should not depend on the market value API call—they should function independently.

Reply:
This type confused me:
```ts
type BaseTokenBalance = {
  balance: string
  balanceUsd: string
  convertedLast: {
    [key in BaseTokenConvertEnum]: string
  }
}
```

Reply:
You are right, probably we need update it.